addons/test_mail/tests/test_mail_channel.py

```diff
@@ -299,6 +299,42 @@ def test_channel_auto_unsubscribe_archived_or_deleted_users(self):
         self.assertEqual(self.test_channel.channel_partner_ids, self.user_employee.partner_id | test_partner)
         self.assertEqual(test_chat.channel_partner_ids, self.user_employee.partner_id | test_partner)
 
+    def test_channel_unfollow_should_also_unsubscribe_the_partner(self):
+        self.test_channel.message_subscribe(self.test_partner.ids)
+        self.test_channel._action_unfollow(self.test_partner)
+
+        self.assertFalse(self.test_channel.message_partner_ids)
+
+    def test_channel_unfollow_should_not_post_message_if_the_partner_has_been_removed(self):
+        '''
+        When a partner leaves a channel, the system will help post a message under
+        that partner's name in the channel to notify others if `email_sent` is set `False`.
+        The message should only be posted when the partner is still a member of the channel
+        before method `_action_unfollow()` is called.
+        If the partner has been removed earlier, no more messages will be posted
+        evev if `_action_unfollow()` is called again.
+        '''
+        test_user = self.env['res.users'].create({
+            "login": "adam",
+            "name": "Jonas",
+        })
+        self.test_channel.write({'email_send': False})
+        self._join_channel(self.test_channel, test_user.partner_id)
+        messages_0 = self.env['mail.message'].search([('res_id', '=', self.test_channel.id), ('author_id', '=', test_user.partner_id.id)])
+        # no message should be posted under test_user's name
+        self.assertEqual(len(messages_0), 0)
+
+        self.test_channel._action_unfollow(test_user.partner_id)
+        messages_1 = self.env['mail.message'].search([('res_id', '=', self.test_channel.id), ('author_id', '=', test_user.partner_id.id)])
+        # a message should be posted because test_user just unfollowed the channel
+        self.assertEqual(len(messages_1), 1)
+
+        self.test_channel._action_unfollow(test_user.partner_id)
+        messages_2 = self.env['mail.message'].search([('res_id', '=', self.test_channel.id), ('author_id', '=', test_user.partner_id.id)])
```

Comment:
Isn't it possible to use assertNoNotifications tool method in combination to use the mail mock ?